feat: CLI commands `meltano cloud login` and `meltano cloud logout`

cjohnhanson opened the pull request:
Implements `login` and `logout` commands for Meltano Cloud. 

Resolves https://github.com/meltano/meltano/issues/7200

WillDaSilva commented:
@cjohnhanson I've changed the base branch from `main` to `cloud`. Feel free to close and re-create the PR if you'd rather make it clean.

Commit:
Fixing branch

Commit:
Removing irrelevant test

Commit:
poetry.lock

Commit:
Docstrings for callback server app.

Commit:
Add platformdirs to cloud-cli only

Commit:
Add cached-property dep to cloud-cli

Commit:
platformdirs in both pyproject.yamls

Commit:
cached-property version

Commit:
merging cloud run changes

Commit:
Re-add platformdirs

Commit:
Re-add platformdirs to root pyproject.toml

Commit:
Add pytest-structlog

Commit:
Merge branch 'cloud' into 7200-cloud-login

Commit:
Add auth and config tests, use aiohttp

Commit:
add config tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.ci hooks

for more information, see https://pre-commit.ci

Commit:
Update docs

Commit:
Add client/__init__.py to replace client.py

Commit:
Add flask to cloud-cli deps

Commit:
re-import run command

Commit:
Add config path flag

Commit:
docstrings, typing

Commit:
Linting

Commit:
use meltano.cloud.cli version in user-agent header

Commit:
Update cli tests

Commit:
Write new config config is not valid json.

Commit:
Add new app client ID

aaronsteers commented:
@cjohnhanson - does this pr also include the logout capability?

I linked to the issue but want to make sure I'm not messing anything up if this isn't 1:1 with that.

WillDaSilva commented on src/cloud-cli/pyproject.toml:
```diff
@@ -41,6 +41,7 @@ python = ">=3.7,<3.12"
 requests = "^2.28.0"
 structlog = "^22.3.0"
 platformdirs = "^3.0.0"
+flask = "^2.1"
```
@cjohnhanson Why Flask?

cjohnhanson commented:
> @cjohnhanson - does this pr also include the logout capability?
> 
> I linked to the issue but want to make sure I'm not messing anything up if this isn't 1:1 with that.

@aaronsteers yes, this has logout capability as well -- though there will be a one-line Pulumi update that needs to go out to before this, I realized in testing that you have to provide a redirect URI for the logout and we'll need to add it as an allowed URI in Cognito.

Commit:
Use config path as passed in callback server

Commit:
Use provided config for run

Commit:
Pass tmp config in tests

Commit:
Make config file fixture

Commit:
Create parent dirs

Commit:
Cleanup

Commit:
Type checking

Commit:
cleanup

Commit:
Ignore path type for config

edgarrmondragon commented on src/cloud-cli/tests/cli/test_cli.py:
```diff
@@ -1,11 +1,13 @@
 
 
-def test_cli():
+def test_cli(tmp_path: Path):
```
I think this fixture isn't used

Commit:
PR feedback

cjohnhanson commented on pyproject.toml:
```diff
@@ -95,6 +95,8 @@ tzlocal = "^4.2.0"
 virtualenv = "^20.17.1"
 werkzeug = ">=2.1,<=2.1.3"
+platformdirs = "^3.0.0"
+pytest-structlog = "^0.6"
```
Yeah, good catch.

Commit:
Add pytest-structlog to deps in noxfile

Commit:
Update logout, address synch feedback

Commit:
-TODO

Commit:
Correct expected response code for logout

Commit:
Add logout page

Commit:
Merge branch 'cloud' into 7200-cloud-login

WillDaSilva commented on src/cloud-cli/meltano/cloud/README.md:
```diff
@@ -0,0 +1,85 @@
+
+As part of authentication, `meltano-cloud` runs a lightweight, short-lived HTTP
+server on `localhost` to handle authentication callbacks. By default, this
+server runs on port 8080 but can be configured to another port using this setting.
```
Is this out of date? I thought port 9999 was being used? This port number also appears in a couple other spots.

cjohnhanson replied:
yep, that is out of date. Updated. 

The only other place I could find it was in unit tests, which is fine. It can use any port locally but cognito will throw an error if you try to redirect to anything besides 9999 on localhost.

Commit:
Update port in README

WillDaSilva committed:
Merge branch 'cloud' into 7200-cloud-login

aaronsteers commented on src/cloud-cli/meltano/cloud/README.md:
```diff
@@ -0,0 +1,85 @@
+- Command line flag: `--auth-callback-port`
+- Environment variable: `MELTANO_CLOUD_AUTH_CALLBACK_PORT`
+
+#### `base_url`
+
+This is the base URL used for API calls to Meltano Cloud, `https://api.meltano.cloud`
+by default.
+
+- Command line flag: `--base-url`
+- Environment variable: `MELTANO_CLOUD_BASE_URL`
```
Should this have the `internal.` prefix?

```suggestion
This is the base URL used for API calls to Meltano Cloud, `https://internal.api.meltano.cloud`
```

cjohnhanson replied:
Yep, will update.

aaronsteers commented on src/cloud-cli/meltano/cloud/README.md:
```diff
@@ -0,0 +1,85 @@
+
+`meltano-cloud` is a standalone CLI tool for interacting with the Meltano Cloud API.
+
+## Configuration
+
+`meltano-cloud` can be configured via
+`meltano-cloud config set <setting name> <setting value>`.
+
```
I was thinking that we would preserve `meltano-cloud config set` for project settings and secrets that would be stored in the cloud.

Is that still the case, or would this also be used for the below options like `auth_callback_port`?

We need to think a bit about collisions and behavior of get/set if so.

cjohnhanson replied:
That makes sense -- this line in the docs is actually outdated, as we had settled in another discussion on configuring exclusively via env vars and flags, since there are very few cloud-specific config options. I'll remove this here.

aaronsteers commented on src/cloud-cli/meltano/cloud/README.md:
```diff
@@ -0,0 +1,85 @@
+- Command line flag: `--base-auth-url`
+- Environment variable: `MELTANO_CLOUD_BASE_AUTH_URL`
+
+#### `organization_id`
+
+This is the default ID for the Meltano Cloud organization to use in interacting with
+the Meltano Cloud API.
+
+- Command line flag: `--organization-id`
+- Environment variable: `MELTANO_CLOUD_ORGANIZATION_ID`
```
Do we need this to be configurable? 

If not, I prefer to drop it and avoid needing to support permanent user-facing organization IDs, which aren't yet a thing we're supporting/declaring for Cloud users here: https://github.com/meltano/cloud-docs/blob/main/docs/concepts.md

cjohnhanson replied:
I'd like to leave it configurable for ease of testing, but a middle path might be to remove from docs and not explicitly support it.

aaronsteers replied:
That makes sense. Wdyt of renaming with an "internal" prefix like "MELTANO_CLOUD_INTERNAL_ORG_ID"?

In that case, we can worry less about whether it's in the docs or not, since it's more self-documenting as non-user-facing.

edgarrmondragon commented on src/cloud-cli/meltano/cloud/api/client/__init__.py:
```diff
@@ -94,32 +108,56 @@ def session(self) -> ClientSession:
+        Yields:
+            None
+        """
+        if not self.auth.logged_in():
```
```suggestion
        if not await self.auth.logged_in():
```

aaronsteers commented on src/cloud-cli/meltano/cloud/README.md:
```diff
@@ -0,0 +1,85 @@
+- Command line flag: `--organization-id`
+- Environment variable: `MELTANO_CLOUD_ORGANIZATION_ID`
+
+#### `project_id`
+
+For project-specific commands, this is the project ID to use by default in interacting
+with the Meltano Cloud API.
+
+- Command line flag: `--project-id`
+- Environment variable: `MELTANO_CLOUD_PROJECT_ID`
```
This is potentially confusing due to the name collision of Meltano Core `project_id` and the (internal) "Cloud Project ID". I prefer not to have any user-facing references to Cloud Project ID unless their clearly delineated as something different.

Can we remove from user-facing docs or maybe give a new name to this concept?

The obvious suggestion of `cloud_project_id` does result in the unfortunate "`MELTANO_CLOUD_CLOUD_PROJECT_ID`", but if we can work around this, I'd be fine with just calling it `cloud_project_id` - since that is different enough in my mind to still be clear from a variety of contexts.


FWIW, we don't have the ID listed here yet, but we should add it if we are going to make it user-facing: https://github.com/meltano/cloud-docs/blob/main/docs/concepts.md

aaronsteers commented on src/cloud-cli/meltano/cloud/README.md:
```diff
@@ -0,0 +1,85 @@
+
+`meltano-cloud` resolves configuration setting values as follows:
+
+- Use a value provided in a command-line flag, for example
+  `meltano-cloud --project-id <project_id> <subcommand>`.
+- Use a value provided in an environment variable, for example `MELTANO_CLOUD_PROJECT_ID`.
```
Ditto comment above re "project id".

From this context, a user would likely assume their project_id from `meltano.yml` would work here. We probably should do one of the following:

1. Use `--cloud-project-id` to be explicit, OR
2. Take out the reference to project ID, OR
3. Rename to something different.

WillDaSilva commented:
The first time I try to login, I get the following error in my browser (Firefox) console:

```
Scripts may not close windows that were not opened by script.
```

It closes as expected on all other attempts.

After authentication the CLI hangs.

WillDaSilva commented:
@cjohnhanson I've left my review suggestions in another PR:
- https://github.com/meltano/meltano/pull/7380

They're mostly small stuff, but a few things I consider important have change there too (e.g. restricting the file permissions on the config file which contains the secret tokens). I've also addressed all of the lint issues.

Commit:
Update src/cloud-cli/meltano/cloud/api/client/__init__.py

Co-authored-by: Edgar R. M. <edgar@meltano.com>

Commit:
Update src/cloud-cli/meltano/cloud/api/auth/auth.py

Co-authored-by: Edgar R. M. <edgar@meltano.com>

WillDaSilva committed:
chore: Review suggestions for `meltano cloud login` PR (#7380)

* Sort deps in `pyproject.toml`

* Update `poetry.lock`

* Add missing word to Cloud CLI contributing doc

* Remove manual line-wrapping in Cloud CLI `README.md`

* Update CLI endpoint docstrings

These docstrings are used as the help text in the CLI, so they shouldn't abide by our regular docstring conventions.

* Address lint issues

`pre-commit run --all-files` now passes

* Address `mypy` issues

* Update Cloud CLI readme

* Fix tests & type checking

* Remove unused `models` modules

* Make Click perform validation on the config file path

* Apply suggestions from code review

* Use deployment name instead of environment name for `meltano cloud run`

* Revert "Apply suggestions from code review"

This reverts commit c6e9f24866145608b3aa99451af714f8c1d79514.

* Restrict file permissions to secret-containing config file

* Update Cloud CLI doc about token storage

* Finish `environment` -> `deployment` rename

---------

Co-authored-by: Cody J. Hanson <cody@meltano.com>

Commit:
remove unused settings

Commit:
Add close window message for Firefox.

cjohnhanson commented:
> The first time I try to login, I get the following error in my browser (Firefox) console:
> 
> ```
> Scripts may not close windows that were not opened by script.
> ```
> 
> It closes as expected on all other attempts.

@WillDaSilva good callout. I've added a `You may close this window.` message to the login/logout pages for any browsers that won't run that `window.close()` call for whatever reason.

cjohnhanson commented:
@aaronsteers I went ahead and just removed the settings section for the docs. Users shouldn't be configuring any of those configurable options (but we should still keep them around for testing purposes), with the one exception of `MELTANO_CLOUD_PROJECT_ID` which isn't actually used anywhere in this PR. The `meltano cloud run` command has `--project-id` as its own flag, and I think it's reasonable to stick to that pattern in the short term until we settle on a top-level "default" project to use when one isn't provided to subcommands that need it, maybe with an associated env var like `MELTANO_CLOUD_DEFAULT_CLOUD_PROJECT`

Commit:
Update test to check id_token

edgarrmondragon commented on src/cloud-cli/meltano/cloud/api/config.py:
```diff
@@ -0,0 +1,171 @@
+        Returns:
+            A MeltanoCloudConfig
+        """
+        with open(config_path, encoding="utf-8") as config_file:
```
```suggestion
        with Path(config_path).open(encoding="utf-8") as config_file:
```

edgarrmondragon commented on src/cloud-cli/meltano/cloud/api/config.py:
```diff
@@ -0,0 +1,171 @@
+
+    def refresh(self) -> None:
+        """Reread config from config file."""
+        with open(self.config_path, encoding="utf-8") as config_f:
```
```suggestion
        with Path(self.config_path).open(encoding="utf-8") as config_f:
```

edgarrmondragon commented on src/cloud-cli/tests/api/auth/test_callback_server.py:
```diff
@@ -0,0 +1,71 @@
+        assert response.status_code == 204
+        assert config.access_token == "meltano-cloud-testing"  # noqa: S105
+        assert config.id_token == "meltano-cloud-testing"  # noqa: S105
+        with open(config.config_path) as config_file:
```
```suggestion
        with Path(config.config_path).open() as config_file:
```

WillDaSilva replied:
@edgarrmondragon Why prefer `Path.open` over `open`?

edgarrmondragon replied:
Really just a nit but I like the API better, and I guess I've been looking at Ruff-linted code 🙈

WillDaSilva replied:
I see: https://beta.ruff.rs/docs/rules/#flake8-use-pathlib-pth

> `open()` should be replaced by `Path.open()`

I'm not a huge fan of that rule in particular (`PTH123`), but Ruff can automatically make it consistent, and I value consistency, so I'll support it in most cases.

Unfortunately it appears that `Path.open` doesn't support all of the parameters that `open` does. Notably it's missing the `opener` parameter, which is useful for opening a file with non-default permission bits set. We open the Cloud CLI config/secrets file for writing with 0600 as its permissions using that.

edgarrmondragon commented on src/cloud-cli/tests/api/auth/test_callback_server.py:
```diff
@@ -0,0 +1,71 @@
+        config.refresh()
+        assert config.id_token is None
+        assert config.access_token is None
+        with open(config.config_path) as config_file:
```
```suggestion
        with Path(config.config_path).open() as config_file:
```

edgarrmondragon commented on src/cloud-cli/tests/api/test_cloud_config.py:
```diff
@@ -0,0 +1,102 @@
+    @pytest.fixture(scope="function")
+    def config_path(self, tmp_path: Path, config_dict: dict):
+        config_file_path = tmp_path / "meltano-cloud.json"
+        with open(config_file_path, "w") as config_file:
```
```suggestion
        with Path(config_file_path).open("w") as config_file:
```

edgarrmondragon commented on src/cloud-cli/tests/api/test_cloud_config.py:
```diff
@@ -0,0 +1,102 @@
+    ):
+        self.config_assertions(subject)
+        os.remove(config_path)
+        with open(config_path, "w") as config_file:
```
```suggestion
        with Path(config_path).open("w") as config_file:
```

edgarrmondragon commented on src/cloud-cli/tests/api/test_cloud_config.py:
```diff
@@ -0,0 +1,102 @@
+        self.config_assertions(subject)
+        subject.organization_id = "organization-id-changed"
+        subject.write_to_file()
+        with open(subject.config_path) as config_file:
```
```suggestion
        with Path(subject.config_path).open() as config_file:
```

edgarrmondragon reviewed (changes requested):
@cjohnhanson I think the change from `access_token` to `id_token` is causing the auth check to get into an infinite loop. The `oauth2/userInfo` seems to require the `access_token`.

Commit:
Use access token for auth-specific calls.

Commit:
-blank line

Commit:
Update src/cloud-cli/tests/api/test_cloud_config.py

Co-authored-by: Edgar R. M. <edgar@meltano.com>

Commit:
Update src/cloud-cli/tests/api/test_cloud_config.py

Co-authored-by: Edgar R. M. <edgar@meltano.com>

Commit:
Update src/cloud-cli/tests/api/test_cloud_config.py

Co-authored-by: Edgar R. M. <edgar@meltano.com>

Commit:
Update src/cloud-cli/tests/api/auth/test_callback_server.py

Co-authored-by: Edgar R. M. <edgar@meltano.com>

Commit:
Update src/cloud-cli/tests/api/auth/test_callback_server.py

Co-authored-by: Edgar R. M. <edgar@meltano.com>

Commit:
Update src/cloud-cli/meltano/cloud/api/config.py

Co-authored-by: Edgar R. M. <edgar@meltano.com>

Commit:
Update src/cloud-cli/meltano/cloud/api/config.py

Co-authored-by: Edgar R. M. <edgar@meltano.com>

edgarrmondragon reviewed (approved):
It's working! 😄

WillDaSilva reviewed (approved):
:shipit:
